_posts/2020-04-02-whats-the-worst-that-can-happen.md

```diff
@@ -0,0 +1,55 @@
+
+## How We Perform Changes
+
+Obviously there are exceptions to every rule, but mostly actions are taken via the use of Jenkins. A job is defined for many of the common actions we need to take, deploys are all scripted in this way for example. This allows us to perform actions on a large number of servers consistently. If there isn't a specific job in Jenkins to perform an action we have a job that allows us to run a command we define ourselves. When planning a change we write out the steps we will take, and provide evidence that these steps produced the desired result in our test environments. When choosing a Jenkins job to run we choose not only the appropriate job, but also any options we will pass to that job. That plan is then read by another Engineer, generally one of the Senior Engineers, and preferably one uninvolved in the change being planned. If the approving Engineer has any doubts then the change needs to be amended to address those doubts.
+
+## What Was the Plan
+
+So onto our change. The plan was reasonably straightforward.
+- Deploy configuration changes to our production environment in Chef
+- Run chef-client on all the Java application servers to make the configuration changes in a consistent manner
+- Restart the Java applications in sequence to pick up those changes
+
+It is worth explaining that last step in more detail, as that is where things did not go according to plan. As multiple different applications were being restarted the plan did not use the application restart job that we have scripted in Jenkins. This job is designed to restart a single application. Instead the job mentioned earlier to run commands we define was used. It was planned as:
+- Environment: Production
+- Search: All our Java Application Servers
+- Command: Restart Java Application
+- Concurrency: 1
+
+These exact steps had worked in all our other environments without issue.
+
+## What Went Wrong?
+
+As we all know, mistakes happen, people are imperfect beings, that is why we script as much of our work as we can. In this case the Engineer forgot to set the Concurrency on the job to restart the applications. In the Application Restart job, this would have caused a problem, but not a major one, as it always starts with a single server, and prompts the Engineer to check that the application has started cleanly before moving on. But that job was not used, the job that was used was an older job, it had fewer safe guards, and unfortunately it had a default value for the concurrency.
```

Comment:
```suggestion
As we all know, mistakes happen, people are imperfect beings, and that's why we script as much of our work as we can. In this case the Engineer forgot to set the Concurrency on the job to restart the applications. In the Application Restart job, this would have caused a problem, but not a major one, as it always starts with a single server, and prompts the Engineer to check that the application has started cleanly before moving on. But that job was not used; the job that was used was an older job, it had fewer safe guards, and unfortunately it had a default value for the concurrency.
```